scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,191 @@
+* There is a bug fix in ClairCore outside the "versioned scanners"
+  * For example, [ClairCore versions up to v1.5.27 were unable to follow hardlinks in image layers](https://github.com/quay/claircore/pull/1305). This was caught when indexing an image whose base operating system information was found in a hard-linked file. So, indexing prior to this resulted in an unknown operating system. Scanner V4 should re-index affected images.
+* Required mapping data was not available during indexing
+  * For example, it may be possible an image is indexed before it's related repository CPEs were made available [here](https://security.access.redhat.com/data/metrics/repository-to-cpe.json). Once the data is available, the image should be re-indexed to ensure it's complete.
+
+Talking with the Clair team and inspecting the [Quay](https://github.com/quay/quay) and [Clair](https://github.com/quay/clair/blob/v4.7.4/httptransport/indexer_v1.go#L175) codebases, it is clear Clair has a built-in way to handle "versioned scanner" updates:
+it is up to the client (typically Quay) to track the Indexer's state (which is a hash of all the "versioned scanners" and their respective versions) returned by Clair to the client via `etag` alongside the Index Report.
+The client then sends that state back to Clair in an `If-None-Match` header, so it may determine if there have been any updates to the "versioned scanners" since indexing the image.
+When done this way, the client does not need to generate a [`*claircore.Manifest`](https://github.com/quay/claircore/blob/v1.5.29/manifest.go#L5), which may be expensive.
+
+It is also possible to forgo this state check and just request Clair to index the image again. ClairCore will check if an Index Report already exists, which "versioned scanners" were used, and compare them to the current versions.
+ClairCore will then only re-index if there is a difference in "versioned scanner" versions and only with the scanners which updated.
+However, this requires the client to generate a `*claircore.Manifest`, which, again, may be expensive. In StackRox's case, this may be expensive because
+StackRox must reach out to the registry several times to determine the fully resolved layer URIs.
+When a registry does not handle the `Range` header (Scanner V4 Indexer uses `Range: bytes=0-0`), then this is even more expensive, as the entire layer may potentially be sent over by the registry.
+
+There is currently no built-in way for Clair to handle the other two mentioned reasons aside from deleting the Index Report and then requesting Clair to Index the image.
+
+Related: Scanner V4 never deletes Index Reports, even when StackRox Central/Sensor no longer tracks the image, so the database is ever-growing.
```

Comment:
Yeah, I saw that, and we could tell Scanner V4 to delete Index Reports once that's called.

But that has some complications:

* We'd have to tell every Scanner V4 Indexer about it because it's not clear which cluster handled that image
* There are other ways images are deleted (pruned after a week or so of inactivity I think). This also has the concern about telling the right scanner or all scanners
* Something Clair ran into: times when the deletions are not properly forwarded. Do we retry? How would we know to retry?


I think it's just simpler for Scanner V4 to worry about this instead